Added support to build libmono with disabled JIT on Windows.

lateralusX opened the pull request:
Current mono build on (at least Windows) fails if build with disabled JIT since the .def file used when building libmono, mono-2.0-sgen.dll, exports a couple of symbols missing when DISABLE_JIT is defined.

Turns out that these methods have been incorrectly exported and should not be part of the public API surface. This commit removes MONO_API for the methods incorrectly marked:

mono_emit_native_call
mono_inst_name
mono_replace_ins

It also removes them from the .def files so they won't be exported in the mono runtime DLL on Windows.

For mono_set_break_policy an empty stub will be added since it is considered being a public API.

The above support is primarily used on none desktop Windows platforms when building mono runtime module without JIT support.

This commit also fix a problem introduced in tramp-amd64.c by commit 4281399 causing it to get duplicated definitions of mono_arch_get_enter_icall_trampoline when DISABLE_JIT has been defined.

vargaz commented:
Some of the changes look unrelated.

lateralusX commented:
I piggy backed  a couple of related changes on this PR for the reversed use case, building with JIT support on non windows desktop platforms. Since I worked on both use cases and these where the changes needed to solve both, I included them in the same PR. Updated the description to reflect this.

kumpera commented on mono/mini/branch-opts.c:
```diff
@@ -1483,6 +1483,15 @@ mono_optimize_branches (MonoCompile *cfg)
 
 #else /* !DISABLE_JIT */
 
+#ifdef HOST_WIN32
```
Why this is needed only under windows?

kumpera commented on mono/mini/helpers.c:
```diff
@@ -308,6 +308,14 @@ mono_disassemble_code (MonoCompile *cfg, guint8 *code, int size, char *id)
 
+#ifdef HOST_WIN32
+const char*
+mono_inst_name (int op)
```
Why do we need this under windows only?

kumpera commented on mono/mini/method-to-ir.c:
```diff
@@ -14550,6 +14550,21 @@ NOTES
 #else /* !DISABLE_JIT */
 
+#ifdef HOST_WIN32
+gpointer
```
Same thing, why only windows need those?

kumpera commented on mono/mini/exceptions-amd64.c:
```diff
@@ -885,7 +885,7 @@ mono_arch_exceptions_init (void)
 }
 
-#if defined(TARGET_WIN32) && !defined(DISABLE_JIT)
+#if defined(TARGET_WIN32) && !defined(DISABLE_JIT) && defined(MONO_ARCH_HAVE_UNWIND_TABLE)
```
This should only check for MONO_ARCH_HAVE_UNWIND_TABLE as that define is already behind `#if defined(TARGET_WIN32) && !defined(DISABLE_JIT)`

kumpera replied:
BTW, that define should be unified with MONO_ARCH_HAVE_CODE_CHUNK_TRACKING as they are both defined under the same set of conditions.

lateralusX commented:
Fixes was done Windows specific since that's where we have the .def file exporting the symbols from the mono DLL and I wanted to minimize the impact of this change as much as possible. If we believe it make sense for other platforms to have the stubs, I could drop the windows defines and include them on all platforms. This PR was originally a small piece of a much bigger PR around improved unwind info on Windows and since there needs to be some adjustments to this PR I think it could be better to actually include the above changes as a separate commit in the unwind PR and close this PR.

joncham commented:
We use .def files on Windows for two reasons.

First, because long ago Mono didn't track exported functions as non-Windows platforms exported everything by default. Rather than add attributes to public API functions at the time, the decision was to use a def file on Windows. 

Second, even with dllexport on public APIs, "exported" symbols from a static library linked into a dynamic library (libmonoruntime -> libmono) don't get exported: https://blogs.msdn.microsoft.com/oldnewthing/20140321-00/?p=1433

Personally, I would love to kill the .def file usage. The first reason for them no longer exists. The second we could solve by either:

1. Force references to those functions somewhere in libmono build. Maybe by storing function address into an array of function pointers. This is better than def file as at least we can use preprocessor to include/exclude functions.

```
void* ExportedFunctions = { 
&mono_class_from_name,
#ifdef DISABLE_JIT
&mono_inst_name,
#endif
...};
```

2. Adjust VS based build to not use the intermediate static libraries and just compile libmono dynamic libraries with all the source. I like this option less. This is what old VS projects did, and it was hard to keep in sync with Makefile based builds.

joncham commented:
I just verified the first approach could work. I can spin up a PR for that...

lateralusX commented:
Getting away from the .def files is something that I considered doing for a while as well. It would have been excellent if that is something you (@joncham) would like to look into.

vargaz commented:
What is the difference between listing functions in a .def file, and listing them in some function list ? Why can't MSVC keep exported functions, if the unix linker can ?

joncham commented:
See the link in my original comment on how the MSVC linker works:
```
Let's go back to the classical model for linking. Code is pulled from a LIB file only if the linker encounters a reference to it. If the linker encounters no reference to any symbols offered by an OBJ file in that LIB, then that OBJ file is not included in the final image. (Remember, we're talking about OBJ files inside LIBs; explicitly-provided OBJ files are always included in the image, under the classical model.)

Now consider a LIB which contains an OBJ file that has a function marked __decl­spec(dll­export). Suppose that no symbol offered by that OBJ file is ever required by the final image. That means that the OBJ file is never added to the image. And that means that the linker does not see the __decl­spec(dll­export) qualifier on the function inside the OBJ file (since the OBJ file was never used), so the function doesn't get exported.
```

From how I read that, if when we build libmono it's source doesn't explicitly need a symbol from an object file in libmonoruntime archive, then the object file is not considered by the linker and the export symbol is not encountered. The linker treats `lib` files differently than `obj` files.

As for difference between def file and listing functions, the advantage to listing functions is it's compiled including the preprocessor. The def file is a static list of functions that must be present, so we cannot disable/exclude public API or adjust for different configs. For example, we already have a different def file for sgen and boehm.

But maybe that's the root issue. Should public API change based on configuration settings when building? Right now the def file for sgen includes `mono_gc_register_bridge_callbacks` and `mono_gc_wait_for_bridge_processing` while boehm does not.

joncham commented:
Besides all that, it sucks to keep def file in sync with code changes when someone adds/removes public API. Of course a list of functions in source has same drawback.

lateralusX commented:
Updated this PR, removed things not related to the DISABLE_JIT use case that has been included in PR https://github.com/mono/mono/pull/4444. This PR still has some relevance (and now also fix a new DISABLE_JIT error in tramp-amd64.c) I believe we should merge it independent of the .def file changes done in https://github.com/mono/mono/pull/4435.

kumpera commented:
DISABLE_JIT is working on non-windows platforms. What's calling the stub'd functions on windows?

kumpera commented:
Can't we remove those functions from the def files? They are not supposed to be externally callable beyond the JIT itself.

lateralusX commented:
Things now starts to clear up. It turns out that a couple of the methods added as stubs in this commit have been incorrectly marked as MONO_API in the past and therefore exported in the def files as public methods, but in reality they should be kept private. I will change those methods to be private, dropping MONO_API in their signatures and remove them from the def files. This will also remove the need for empty stubs, since they will no longer be exported.

This is the plan:

mono_inst_name -> mark as MONO_LLVM_INTERNAL, remove from def files.
mono_emit_native_call -> move to mini.h, remove MONO_API, remove from def files.
mono_replace_ins -> remove MONO_API, remove from def files.
mono_break_point_policy -> should be public and will need empty stub method.

lateralusX commented:
Updated with changes according to above plan.

vargaz commented on mono/mini/decompose.c:
```diff
@@ -20,7 +20,6 @@
 #ifndef DISABLE_JIT
 
 /* FIXME: This conflicts with the definition in mini.c, so it cannot be moved to mini.h */
-MONO_API MonoInst* mono_emit_native_call (MonoCompile *cfg, gconstpointer func, MonoMethodSignature *sig, MonoInst **args);
```
The fixme should be removed too.

lateralusX replied:
OK, will remove the fixme comment. Thought it was related to all three functions since they where grouped together.

vargaz commented:
Looks ok otherwise.

lateralusX commented:
Removed fixme comment.

kumpera commented:
LGTM
